README
-------------------------------------------------------------------------------
Welcome to ixlib.
-------------------------------------------------------------------------------

Table of contents:
1. What is ixlib?
2. Installation
3. Implementation notes
4. FAQ
5. Documentation
6. TODO

URL:
	http://ixlib.sourceforge.net
Authors:
	Andreas Kloeckner <ak@ixion.net>, Hardy Kahl <hk@ixion.net>

If you use ixlib in your project, I would love to hear from you. I
might even advertise your project on the ixlib homepage. :-))) ixlib
is the result of what I needed in my projects and was able to abstract
out without too much trouble. Though it is quite a bag of mixed stuff,
I'm sure you'll like at least one or two components.

Its earliest roots go back to a library Hardy and I wrote for Watcom
C++ under DOS. Right now, it compiles fine under Linux, Win32 and 
FreeBSD.

Have fun!
Andreas

ATTENTION! In versions 0.95.0 and 0.96.0, there have been incompatible
changes to the JavaScript interface! See the ChangeLog for details.
(Those changes were necessitated by major feature improvements, however.)

-------------------------------------------------------------------------------
1. What is ixlib?
-------------------------------------------------------------------------------

ixlib is a small c++ tools library based upon the standard template
library.  It provides

* a javascript interpreter (subset of ECMAscript 4, strict mode) 
  [ixlib_javascript.hh]
* an exception handling framework [ixlib_exbase.hh]
* garbage collection [ixlib_garbage.hh]
* automatic array management [ixlib_array.hh]
* planar geometry (rectangles, regions) [ixlib_geometry.hh]
  polygons (rasterization, convex hull, smoothing, removal of crossings) 
  [ixlib_polygon.hh]
* rasterization [ixlib_drawing_functions.hh]
* matrices (including linear system solver, Cholesky and LU decomposition,
  determinants, inversion, Gauss and Gauss-Jordan elimination)
  [ixlib_matrix.hh]
* command line parsing [ixlib_cmdline.hh]
* versatile int <-> string conversions [ixlib_numconv.hh]
* regular expressions [ixlib_re.hh]
* XML parsing (non-DTD) [ixlib_xml.hh]

Some of the guidelines I tried to follow are:

* use a separate namespace "ixion"
* try to be STL-look-alike
* no unnecessary dependencies: you pay only for those parts that you use
  (pay means: use disk space, take execution time)
* no unnecessary dependencies 2: The library doesn't do any i/o by itself,
  except if given a stream to explicitly do so. Besides flex and the STL,
  there are no other dependencies.
* do not use abbreviations
* conform to standards (XML, ECMAscript)
* do not duplicate or mimic STL functionality (e.g. no separate
  string class), that is be strictly an STL-add-on
* provide a complete internationalization framework for localized
  texts and error messages

Furthermore, every component of ixlib has been thoroughly tested and
is considered production-quality code.

-------------------------------------------------------------------------------
2. Installation
-------------------------------------------------------------------------------

The usual

configure [--disable-static] [--enable-debug]
make
make check
make install

should do the trick. You can skip the "make check" parts if you do not
want to run the regression tests.

ixlib currently depends on
* a good c++ compiler (gcc 2.95 will do)
* the SGI STL (i.e. an STL implementation containing a "hash_map")
  (normally included with your c++ compiler, e.g. gcc)
* flex

-------------------------------------------------------------------------------
3. Implementation notes
-------------------------------------------------------------------------------

Using ixlib is as easy as abc. To compile, just type

c++ -c my_program.cc `ixlib-config --cflags`

To link, type

c++ my_program.o -o my_program `ixlib-config --libs`

One word on the usage of the templates. In order to avoid incurring the
overhead of compiling/inlining every included template into every source file,
the following scheme is used: The declaration goes into a ixlib_*.hh
file, and the corresponding implementation goes into ixlib_*_impl.hh.
This implementation does not inline by default, so you will use the following
in *one* of your c++ files for every type you desire instantiation for:

#include <ixlib_foo_impl.hh>

template foo_class<int>; // explicit instation for ints

-------------------------------------------------------------------------------
4. FAQ
-------------------------------------------------------------------------------
Q: Does ixlib work with gcc 3.0?
A: Yes.

Q: Does it work with older gccs?
A: It should, down to 2.95.

Q: How can I find out about the usage of a certain component?
A: Just try the appropriate test in the test/ directory. (e.g.
testjs.cc for the javascript interpreter)

Q: Hey, you seem to have forgotten all of the implementation?
A: No I haven't. Look in ixlib_*_impl.hh. See also (3).

Q: What are the speed characteristics of the JavaScript interpreter?
A: If it is compiled with -O -finline-functions, it is about 5 times 
slower than Python, and about 10 times slower than Perl. It seems like
a good match for Netscape's JS implementation. Furthermore, on my machine
it's almost exactly 81 times faster than a C-64 executing the analogous
code. :-)

All benchmarks were taken with code anologous to this one:

var i = 0;
while (i < 100000) ++i;

Leaving out the optimization approximately doubles the execution time.
Also, ++i is a bit faster than i++, for internal reasons.

Q: How can I find out what syntax variants the javascript interpreter
supports without digging into the source?
A: Look into test/test.js. Almost every construct that is supported
is exercised in that file.

Q: How do I create an ixlib release?
A: [This is mainly my own checklist for doing it, you're not really
supposed to do ixlib releases, at least, not yet :-]

- enter release into changelog
- cvs commit
- make dist
- test package on other computer
- test package on win32-cross build
- enter news item on web page (web/ directory)
- make update-web
- enter news item on sourceforge
- cvs tag RELEASE_0_95_0
- do sourceforge release of package
- do sourceforge release of documentation
- do sourceforge release notification
- upload package and lsm to ftp://ftp.ibiblio.org/incoming/Linux
- post to announcement list
- post on freshmeat

Q: I get an error compiling lex.xml.cc or lex.javascript.cc. What's up?
A: Do the following

cd src
ed lex.xml.cc < flex-gcc3-patch.ed
ed lex.javascript.cc < flex-gcc3-patch.ed
cd ..
make

That should do the trick. (older flex-generated files break with gcc3)
Another solution would be to upgrade your flex package. The one from
Debian woody works for me.

-------------------------------------------------------------------------------
5. Documentation
-------------------------------------------------------------------------------

ixlib now has (doxygen-generated) documentation. It can be built via
"make documentation" in the package root directory, and can the be
found the "doc" directory.

If you don't want to install doxygen, you can get a documentation tarball
or view the same HTML documentation on-line at the ixlib homepage.

-------------------------------------------------------------------------------
6. TODO
-------------------------------------------------------------------------------
* ref<const T> isn't really feasible/interoperable with ref<T>

* convert matrix to refs.

* csg on polygons

ChangeLog
2001-10-24  Andreas Kloeckner  <ak@ixion.net>

	* released 0.96.2 to prepare for encryption and bigint functions

2001-10-23  Andreas Kloeckner  <ak@ixion.net>

	* javascript: fixed js_string.split, Array.join 
          (Thanks to Martin Oberzalek <kingleo@gmx.at>)

2001-10-18  Andreas Kloeckner  <ak@ixion.net>

	* javascript: added js_string.split, js_array::push_back,
	  test cases for split (Thanks to Martin Oberzalek <kingleo@gmx.at>)

2001-10-04  Andreas Kloeckner  <ak@ixion.net>

	* src/js_declaration.cc: fixed variable instatiation bug,
	  result of introduction of eliminateWrappers()

2001-10-03  Andreas Kloeckner  <ak@ixion.net>

	* js: fixed class scoping, so that lexical scope variables don't
          shadow instance variables

	* js: fixed makeValue("string") by
	  removing makeValue(bool); (breaks binary compatibility)

	* bumped shlib version to 6

	* js: added traps for division by zero

2001-09-15  Andreas Kloeckner  <ak@ixion.net>

	* added FreeBSD support (libintl fixes,math.h) 

2001-07-31  Andreas Kloeckner  <ak@ixion.net>

	* src/js_declaration.cc: Silly me - constructorless classes would
	  segfault the interpreter. Fixed.

2001-07-17  Andreas Kloeckner  <ak@ixion.net>

	* bumped version to 0.96.2, iface version to 5.
	
	* javascript: separated wrapper elimination and duplication,
	  that functionality is now in two separate methods.

2001-07-06  Andreas Kloeckner  <ak@ixion.net>

	* released version 0.96.1 without big announcements - done to
	  ease SDLucid 0.90.0 release.

2001-07-05  Andreas Kloeckner  <ak@ixion.net>

	* src/ixlib_matrix_impl.hh: implemented matrix norm,
	  more get***() to ***() renaming

2001-07-01  Andreas Kloeckner  <ak@ixion.net>

	* configure.in: bumped version to 0.96.1

	* src/ixlib_geometry.hh,ixlib_xml.hh: 
	  changed names from get***() to ***(), deprecated the old versions.

2001-06-29  Andreas Kloeckner  <ak@ixion.net>

	* released version 0.96.0
	* src/ixlib_random.hh: replaced fabs by NUM_ABS because of gcc3 bug

2001-06-27  Andreas Kloeckner  <ak@ixion.net>

	* src/ixlib_geometry.hh: added coord_vector::op!=

2001-06-25  Andreas Kloeckner  <ak@ixion.net>

	* src/ixlib_polygon.hh: fixed gcc3 name lookup weirdness

2001-06-20  Andreas Kloeckner  <ak@ixion.net>

	* src/Makefile.am: added an ed script patch to fix the broken
	  flex-generated code. (kludge, but works)

	* made the whole library compile with gcc3.

2001-06-16  Andreas Kloeckner  <ak@ixion.net>

	* src/ixlib_geometry.hh: added region::size() call
	  added rectangle::getWidth/Height/Area calls,
          marked getSizeX/Y obsolete

2001-06-10  Andreas Kloeckner  <ak@ixion.net>

	* src/ixlib_numconv.hh: made float2dec behave more predictably

	* src/js_value.cc: fixed string conversion of integers

2001-06-09  Andreas Kloeckner  <ak@ixion.net>

	* js: constructors and scoping are *much* faster and more
	  compliant
	
	* src/ixlib_javascript.hh: removed context parameter for
	  call, callAsMethod and construct. They aren't needed.
          Further, removed the compatibility wrappers for
          call, construct and duplicate in value.
		  
	  !! BE CAREFUL !!
          If you have overloaded call(), construct() or duplicate() in any of
	  your javascript::value implementations, *PLEASE* double-check that 
	  their prototypes match the new ones in ixlib_javascript.hh!
          Otherwise, you're in for some really hard-to-track errors.

2001-06-06  Andreas Kloeckner  <ak@ixion.net>

	* src/ixlib_exgen.hh: removed memory related exceptions, they
	  are superceded by ANSI C++ bad_alloc.

2001-06-01  Andreas Kloeckner  <ak@ixion.net>

	* src/js_value.cc: added string.toString()

2001-05-26  Andreas Kloeckner  <ak@ixion.net>

	* src/lex.*.yy: made both scanners batch scanners for a *HUGE*
	  performance benefit

	* src/ixlib_string.hh: added encoding and decoding to/from base64
	  (as of rfc2045)

2001-05-20  Andreas Kloeckner  <ak@ixion.net>

	* src/ixlib_matrix_impl.hh: normalize() variable thinko corrected,
	  thanks to Vaughn Cato <vaughn@giantstudios.com>

2001-05-09  Andreas Kloeckner  <ak@ixion.net>

	* javascript: improved error reporting

2001-05-04  Andreas Kloeckner  <ak@ixion.net>

	* src/js_value.cc: finally fixed a == null, a != null

	* src/ixlib_javascript.hh: jumbled constness of value methods a bit
          Backward-compatibility wrappers are in place to avoid major 
	  source incompatibility. Users are encouraged to switch to the new
	  non-const interfaces.

2001-05-03  Andreas Kloeckner  <ak@ixion.net>

	* configure.in: bumped version number to 0.96.0

2001-05-01  Andreas Kloeckner  <ak@ixion.net>

	* javascript: added support for classes

2001-04-17  Andreas Kloeckner  <ak@ixion.net>

	* Released version 0.95.1

	* configure.in (ix_ixlib_micro_version): bumped version to 0.95.1

	* src/re.cc: fixed range parsing, added test cases

	* src/ixlib_re.hh: redesign, regex_string is no more a string
	  subclass (which is nonsense, if you think about it).
          It is now subclass to a more generic regex matcher that 
          is a template and can be used outside the domain of strings.
          !! This breaks binary compatibility !!
          !! This might break source compatibility in some rare and
             screwed-up cases !!

2001-03-24  Andreas Kloeckner  <ak@ixion.net>

	* src/js_value.cc: added toFloat to float type and toInt to int type

2001-03-23  Andreas Kloeckner  <ak@ixion.net>

	* Released version 0.95.0
	
	* src/base.cc: disabled i18n for win32

	* src/js_value.cc: function parameters were involuntarily passed by
	  reference, fixed

	* src/js_value.cc: added toInt to float and toFloat to int types

        * src/js_value.cc: 1/5.0 works correctly now, returning 0.2 instead
	  of 0.

	* src/js_interpreter.cc: removed no_free_ref to RootScope,
	  eliminating deep binding issues related to that.
	  !!! ATTENTION !!! This also breaks source compatibility.
          interpreter.RootScope is now a ref!


po/cat-id-tbl.c
/* Automatically generated by po2tbl.sed from ixlib.pot.  */

#if HAVE_CONFIG_H
# include <config.h>
#endif

#include "libgettext.h"

const struct _msg_ent _msg_tbl[] = {
  {"", 1},
  {"Unable to evaluate expression", 2},
  {"Function not yet implemented", 3},
  {"General error", 4},
  {"NULL pointer encountered", 5},
  {"Invalid parameter", 6},
  {"Index out of range", 7},
  {"Buffer overrun", 8},
  {"Buffer underrun", 9},
  {"Item not found", 10},
  {"Invalid operation", 11},
  {"Dimension mismatch", 12},
  {"Operation cancelled", 13},
  {"Unable to operate on empty set", 14},
  {"Unable to remove GC entry", 15},
  {"Unable to protect non-freeable entry", 16},
  {"Unmatched wildcard", 17},
  {"Parameter error: wildcard", 18},
  {"Path syntax error", 19},
  {"Path relative or invalid", 20},
  {"Seek error", 21},
  {"Read request across EOF", 22},
  {"Write request across EOF", 23},
  {"Write protection violation", 24},
  {"File clipping violation", 25},
  {"Internal error", 26},
  {"No such file or directory", 27},
  {"Access denied", 28},
  {"File exists", 29},
  {"Too many open files", 30},
  {"Out of media space", 31},
  {"File too big", 32},
  {"Hardware failure", 33},
  {"Path inaccessible", 34},
  {"Continuity violation", 35},
  {"Not supported", 36},
  {"Invalid file format", 37},
  {"Invalid replacement sequence", 38},
  {"Invalid chunk", 39},
  {"CRC error", 40},
  {"instead of ", 41},
  {"Unterminated comment", 42},
  {"Cannot convert", 43},
  {"Unexpected token encountered", 44},
  {"Unexpected end of file", 45},
  {"Cannot modify rvalue", 46},
  {"Unknown identifier", 47},
  {"Unknown operator", 48},
  {"Invalid non-local exit", 49},
  {"Invalid number of arguments", 50},
  {"Invalid token encountered", 51},
  {"Cannot redeclare identifier", 52},
  {"Constructor called on constructed object", 53},
  {"No superclass available", 54},
  {"'(' in switch statement", 55},
  {"')' in switch statement", 56},
  {"'{' in switch statement", 57},
  {"':' in case label", 58},
  {"':' in default label", 59},
  {"'}' in switch statement", 60},
  {"variable identifier", 61},
  {"constant identifier", 62},
  {"initializer for constant", 63},
  {"function identifier", 64},
  {"'(' in function declaration", 65},
  {"parameter identifier", 66},
  {"')' in function declaration", 67},
  {"'{' in function definition", 68},
  {"method identifier", 69},
  {"'(' in method declaration", 70},
  {"')' in method declaration", 71},
  {"'{' in method definition", 72},
  {"constructor identifier", 73},
  {"class identifier", 74},
  {"'(' in constructor declaration", 75},
  {"')' in constructor declaration", 76},
  {"'{' in constructor definition", 77},
  {"'{' in class declaration", 78},
  {"'function' keyword", 79},
  {"'(' in if statement", 80},
  {"')' in if statement", 81},
  {"'(' in while statement", 82},
  {"')' in while statement", 83},
  {"'while' in do-while", 84},
  {"'(' in do-while statement", 85},
  {"')' in do-while statement", 86},
  {"'(' in for statement", 87},
  {"')' in for statement", 88},
  {"';' in for statement", 89},
  {"break label", 90},
  {"continue label", 91},
  {"'(' in 'new' expression", 92},
  {"')' in 'new' expression", 93},
  {"')' in function call", 94},
  {"']' in subscript", 95},
  {"-> string", 96},
  {" -> int", 97},
  {" -> float", 98},
  {" -> bool", 99},
  {": duplication", 100},
  {": lookup of ", 101},
  {": subscript", 102},
  {": call", 103},
  {": call as method", 104},
  {": construction", 105},
  {": assignment", 106},
  {": operator ", 107},
  {" on ", 108},
  {"prefix ++", 109},
  {"postfix ++", 110},
  {"prefix --", 111},
  {"unary +", 112},
  {"unary -", 113},
  {"undefined", 114},
  {"null", 115},
  {"integer", 116},
  {"floating point", 117},
  {"string", 118},
  {"function", 119},
  {"object", 120},
  {"built-in object", 121},
  {"host object", 122},
  {"scope", 123},
  {"bound method", 124},
  {"type", 125},
  {"unknown value type", 126},
  {"by assignment", 127},
  {"Invalid quantifier", 128},
  {"Unbalanced backreference", 129},
  {"Invalid escape sequence", 130},
  {"Invalid backreference", 131},
  {"Unterminated character class", 132},
  {"Unable to match without expression", 133},
  {"Unknown token", 134},
  {"End of input", 135},
  {"Syntax error", 136},
  {"No text expected in this context", 137},
  {"No literal expected in this context", 138},
  {"Literal expected", 139},
  {"Unexpected end of input", 140},
  {"Close tag expected", 141},
  {"Close tag has different name", 142},
};

int _msg_tbl_length = 142;
